src/dag.rs

```diff
@@ -352,6 +352,12 @@ pub struct SagaDag {
     pub(crate) end_node: NodeIndex,
 }
 
+// The number of nodes for the "start" and "end" nodes in the saga.
+//
+// This constant helps callers iterate over the saga DAG while safely
+// ignoring these indices.
+const START_AND_END_NODES: usize = 2;
```

Comment:
I'm a little worried this abstraction exposes implementation details that could reasonably change in the future.  Dumb examples: we don't _have_ to have "start" and "end" nodes, and we could choose to insert other kinds of internal nodes.  

What about providing an iterator over the nodes instead?  We could have the iterator skip various internal nodes if we want (e.g., "start", "end").

Reply:
To be clear, this constant is not `pub` - I just wanted to be really explicit about the magic "2" number.

If you'd prefer for me to expose an iterator, I can, but it seems kinda equivalent to say "you can access indices [0, N)" (where "N" is what we expose for node count") vs "here is an iterator that returns the integers [0, N)". Either way, the thing exposed in the `pub` interface *does not* expose the start / end nodes.

Reply:
Good call about the `const` not being `pub`.  That does help.

It still seems a little brittle to me.  I think this change creates an implicit dependency that `saga_dag.start_node == saga_dag.graph.node_count() - 2` and `saga_dag.end_node == saga_dag.graph.node_count() - 1`, as well as that there are no other internal nodes that we might want to hide from users.  Previously we've avoided the assumption about the node indexes -- that's why `saga_dag` have `start_node` and `end_node` fields (rather than assuming what indexes those nodes would be).  The indexes these nodes get depends on what order you add them to the graph, which I think changed relatively recently.

Relatedly, I suspect we probably should hide `SubsagaStart` nodes from users: they're not actions and they don't produce any output (and I don't think it makes sense for them to fail).  This has the advantage that the `name` wouldn't need to be an `Option`.  All of the user-visible node types have names.

It feels like an iterator would be a lot cleaner but maybe I'm missing a downside?

Reply:
Okay, in https://github.com/oxidecomputer/steno/pull/67/commits/cf98df1cf0eec3bd5021c10f08284f3120906436 I implemented this as an iterator and updated the test. LMK what you think!